### PR TITLE
:globe_with_meridians: Implement k means algorithm

### LINKED ISSUE #5: :globe_with_meridians: Implement k-means algorithm
# :white_check_mark: What's needed
You must implement the main algorithm of the project.

This algorithm is the `k-means`. It consists of regrouping pixels by k *(given in the file)* clusters according to the color differences.
Until the clusters movements are less than the given convergence, the program must compute new bunch of k clusters and add data to its nearest clusters.

The computation works with color distance *(so distance between color vectors)*. Here's a simple equation to compute it:

image

*Convergence is considered reached when all the clusters have moved less than the convergence parameter since the last
iteration.*

## 📋 Subtasks
- [x] Build the convergence loop
- [x] Make the creation of the clusters 
- [x] Compute the color distance for each iteration
- [x] Display the asked output *(see below)* at the end of the convergence loop

image

The output is formatted like this:
```
--
Cluster color vector1 -> (r,g,b)
-
Cluster linked color vector1 -> (r,g,b)
Cluster linked color vector2 -> (r,g,b)
Cluster linked color vector3 -> (r,g,b)
--
Cluster color vector2 -> (r,g,b)
-
...
```
## :file_folder: Files required
*List all required files if any*

### PR DESCRIPTION
**Pull request by landryarki**
k-means implemented and display the output of all cluster as requested.
the algorithm should handle endless vector as long as they are all the same size.

### CODE REVIEW

File: src/ImageCompressor.hs [hunk 1]
```diff
@@ -5,12 +5,125 @@
...
+recenter [] = []
+recenter ((CentroidBind _ l):xs) =
+  (CentroidBind (getCenter (map snd l)) l) : recenter xs
+
+compareCentroids :: [CentroidBind] -> [CentroidBind] -> Double -> Bool
+compareCentroids [] [] _ = True
+compareCentroids [] _ _ = False
+compareCentroids _ [] _ = False
+compareCentroids ((CentroidBind c1 _):xs) ((CentroidBind c2 _):ys) conv =
+  if getDistance c1 c2 < conv then compareCentroids xs ys conv else False
```
Comment (Thomaltarix) [inline src/ImageCompressor.hs]: You can use guard close here

File: src/ImageCompressor.hs [hunk 2]
```diff
@@ -5,12 +5,125 @@
...
+
+getCentroidFromBindList :: [CentroidBind] -> [Vector]
+getCentroidFromBindList [] = []
+getCentroidFromBindList ((CentroidBind a _):xs) =
+  a : getCentroidFromBindList xs
+
+bindVectorToCentroid :: (Vector, Vector) -> [CentroidBind] -> [CentroidBind]
+bindVectorToCentroid _ [] = []
+bindVectorToCentroid (co, v) ((CentroidBind c l):xs) =
+  if getClosestVector v (getCentroidFromBindList ((CentroidBind c l):xs)) == c
```
Comment (Thomaltarix) [inline src/ImageCompressor.hs]: You can use guard close here

File: src/ImageCompressor.hs [hunk 3]
```diff
@@ -5,12 +5,125 @@
...
+    doImageCompressor (Conf lc) opt [emptyCentroid c | c <- recentered]
+
+putVector :: Vector -> IO ()
+putVector (Vector []) = return ()
+putVector (Vector l) =
+  putStr "(" >>
+  putVector' l >>
+  putStr ")"
+  where
+    putVector' [] = return ()
```
Comment (Thomaltarix) [inline src/ImageCompressor.hs]: You could maybe define another function to avoid defining a function into another

**Review by Thomaltarix — APPROVED** [review]
You can try to rename your variables in a better way if you have time to do it. If you don't get it, it's fine. You can also split your display and your algorithm into different files if you have time too

File: src/ImageCompressor.hs (hunk 3 above)
Comment (landryarki) [reply src/ImageCompressor.hs]: i prefer it that way since putVector' has the sole purpose to be used by and only by putVector

**Review by Thomaltarix — APPROVED** [review]
Thank's for the changes :wink:

**Pull request merged**